File: apps/windows/TeslaSync.App/feature-views/AccordionSection.Model.cs
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The render-time data model the <c>AccordionSection</c> view binds to — the native analogue of the web
/// <c>AccordionSectionProps</c> (web/src/features/system/components/status/AccordionSection.tsx). The web source is a
/// pure presentational disclosure: it takes an already-resolved <c>icon</c> / <c>title</c> / <c>description</c> /
/// <c>defaultOpen</c> (plus the <c>badges</c> and <c>children</c> render nodes, which are supplied to the view as
/// WinUI elements and so are NOT part of this pure-data model) and performs no fetching. The web <c>icon:
/// ReactNode</c> becomes an optional Segoe Fluent <see cref="IconGlyph"/> (the parent supplies a glyph, exactly as it
/// supplies a Lucide node on the web). Because the component has no fetch lifecycle there is no loading / error /
/// stale / offline branch to model — the only states are collapsed and expanded (resolved from
/// <see cref="DefaultOpen"/>), plus a friendly empty caption when an expanded section has no body. Pure data — no
/// WinUI types — so the projection is unit-tested without a UI host.
/// </summary>
/// <param name="Title">The section title shown in the header (web <c>title</c>).</param>
/// <param name="Description">The muted sub-line under the title (web <c>description</c>).</param>
/// <param name="IconGlyph">Optional Segoe Fluent glyph rendered before the title (web <c>icon</c>), or null.</param>
/// <param name="DefaultOpen">Whether the section starts expanded (web <c>defaultOpen</c>, default false).</param>
public sealed record AccordionSectionModel(
    string Title,
    string Description,
    string? IconGlyph = null,
    bool DefaultOpen = false);

/// <summary>
/// The fully projected, render-ready view of one <c>AccordionSection</c> input — the native analogue of everything
/// the web component resolves before returning JSX. Holds the optional <see cref="IconGlyph"/>, the verbatim
/// <see cref="Title"/> + <see cref="Description"/> (with presence flags so a blank line is collapsed rather than
/// rendered as empty space), the initial <see cref="DefaultOpen"/> state, the shared empty-body caption, and the
/// composed header <see cref="AutomationName"/>. Pure data so every field is asserted headlessly.
/// </summary>
/// <param name="IconGlyph">The Segoe Fluent header glyph, or null when the web <c>icon</c> was absent.</param>
/// <param name="HasIcon">Whether a glyph is present.</param>
/// <param name="Title">The section title, rendered verbatim (web <c>title</c>).</param>
/// <param name="HasTitle">Whether a non-blank title is present.</param>
/// <param name="Description">The muted sub-line, rendered verbatim (web <c>description</c>).</param>
/// <param name="HasDescription">Whether a non-blank description is present.</param>
/// <param name="DefaultOpen">Whether the section starts expanded (web <c>defaultOpen</c>).</param>
/// <param name="EmptyMessage">The localized caption shown when an expanded section has no body (never a blank box).</param>
/// <param name="AutomationName">The composed Narrator name for the disclosure header (title then description).</param>
public sealed record AccordionSectionDisplay(
    string? IconGlyph,
    bool HasIcon,
    string Title,
    bool HasTitle,
    string Description,
    bool HasDescription,
    bool DefaultOpen,
    string EmptyMessage,
    string AutomationName);

/// <summary>
/// Pure projection from an <see cref="AccordionSectionModel"/> to its <see cref="AccordionSectionDisplay"/> — the
/// native port of web/src/features/system/components/status/AccordionSection.tsx. The title and description are
/// rendered verbatim (the web interpolates the resolved props unchanged); an empty glyph collapses to null so no
/// icon slot is reserved; the empty-body caption resolves through the i18n facade; and the header Narrator name is
/// composed from the present header parts (title then description) — the expanded / collapsed state itself is spoken
/// by the disclosure control, so it is not duplicated into the name. No WinUI types — unit-tested without a UI host.
/// </summary>
public static class AccordionSectionProjection
{
    /// <summary>i18n key for the empty-body caption shown when an expanded section has no content.</summary>
    public const string EmptyMessageKey = "accordion.empty";

    /// <summary>English fallback for <see cref="EmptyMessageKey"/>.</summary>
    public const string EmptyMessageFallback = "No content";

    /// <summary>Project <paramref name="model"/> into a render-ready display using the i18n facade.</summary>
    /// <param name="model">The render-time data model (the web props minus the badge / body render nodes).</param>
    /// <param name="localizer">The i18n facade the empty-body caption resolves through.</param>
    public static AccordionSectionDisplay Project(AccordionSectionModel model, ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentNullException.ThrowIfNull(localizer);

        string title = model.Title ?? string.Empty;
        string description = model.Description ?? string.Empty;
        string? icon = string.IsNullOrEmpty(model.IconGlyph) ? null : model.IconGlyph;
        bool hasTitle = !string.IsNullOrWhiteSpace(title);
        bool hasDescription = !string.IsNullOrWhiteSpace(description);
        string emptyMessage = localizer.GetString(EmptyMessageKey, EmptyMessageFallback);

        return new AccordionSectionDisplay(
            IconGlyph: icon,
            HasIcon: icon is not null,
            Title: title,
            HasTitle: hasTitle,
            Description: description,
            HasDescription: hasDescription,
            DefaultOpen: model.DefaultOpen,
            EmptyMessage: emptyMessage,
            AutomationName: BuildAutomationName(hasTitle, title, hasDescription, description, emptyMessage));
    }

    private static string BuildAutomationName(
        bool hasTitle,
        string title,
        bool hasDescription,
        string description,
        string emptyMessage)
    {
        // Reading order matches the web header (title then description). Only present parts are spoken so the
        // Narrator name never carries a dangling separator; a fully blank header falls back to the empty caption so
        // the disclosure always has a meaningful accessible name.
        var parts = new List<string>(2);
        if (hasTitle)
        {
            parts.Add(title);
        }

        if (hasDescription)
        {
            parts.Add(description);
        }

        return parts.Count > 0 ? string.Join(". ", parts) : emptyMessage;
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>AccordionSection</c> surface (P1/S11 diagnostics contract). Records only the
/// operational <c>view.opened</c> event with the surface slug — never the title, description or body — so a
/// diagnostics line can never leak fleet data. Thread-safe.
/// </summary>
public sealed class AccordionSectionDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">The sink the <c>view.opened</c> line is written to, or null.</param>
    public AccordionSectionDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=AccordionSection</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={AccordionSectionRegistration.Slug}");
    }
}

/// <summary>
/// Canonical metadata for the <c>AccordionSection</c> feature surface — the native mirror of the web component at
/// <c>web/src/features/system/components/status/AccordionSection.tsx</c>. Holds the diagnostics slug emitted with the
/// <c>view.opened</c> event. UI-free so the metadata is asserted in tests.
/// </summary>
public static class AccordionSectionRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event.</summary>
    public const string Slug = "AccordionSection";
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/AccordionSection/AccordionSectionModel.kt
// Pure, framework-free model + projection for the AccordionSection feature view — the native analogue of
// the data the web component derives from its props before returning JSX
// (web/src/features/system/components/status/AccordionSection.tsx). No Compose, no Android, no HTTP: every
// type here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the composable a
// thin render layer.
//
// The web component is a PURELY PRESENTATIONAL disclosure primitive from the shared `@/components/ui` +
// `@/components/motion` library. It binds NO data hook (the P3 prompt's own "Data sources: none - pure
// presentational component"); its only collaborators are GlassPanel, FadeIn, the cn() class helper, and a
// chevron glyph. The single piece of logic it owns is the open/closed toggle (`setOpen(prev => !prev)`,
// reachable by click and by the Enter/Space key handler) plus the chevron's `open && 'rotate-180'`
// transform; everything else is layout and caller-supplied slots (icon, title, description, badges,
// children). This file reproduces that logic in [AccordionSectionModel] and reduces the surface to its
// genuine, reachable render states in [AccordionRender]:
//
//   • Collapsed       — open == false: only the header is shown (web: `{open && (...)}` is false).
//   • ExpandedContent — open == true with a body slot: header + divider + the caller's content (web FadeIn).
//   • ExpandedEmpty   — open == true with no body slot: header + divider + a friendly empty state. The web
//                       `children` is a required slot, but a native caller may hand none; rather than render
//                       a blank box we show the shared feedback EmptyState, exactly as the sibling CodeBlock
//                       surface does for an empty body.
//
// There is no remote feed behind an AccordionSection, so there is NO loading / error / stale / offline
// lifecycle to model — inventing one would be fabricated state the web source never has (a "No silent
// drift" covenant violation), exactly as the sibling CodeBlock / JsonFormatter surfaces document. The data
// the P3 states checklist describes ("initial fetch", "fetch failed", "older than freshness window", "no
// connectivity") presupposes a data source this component does not have.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/AccordionSection — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path — exactly as the sibling feature-view surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.accordionsection

/** Diagnostics + registry identifiers for the surface (P1/S11). */
object AccordionSectionRegistration {
    /** Stable surface id. */
    const val ID: String = "accordion-section"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "AccordionSection"
}

/** Chevron rotation, in degrees, when the section is open — the web `open && 'rotate-180'` transform. */
const val CHEVRON_OPEN_DEGREES: Float = 180f

/** Chevron rotation, in degrees, when the section is closed (the resting glyph orientation). */
const val CHEVRON_CLOSED_DEGREES: Float = 0f

/**
 * Native-only microcopy defaults. The web AccordionSection owns no i18n keys of its own — its `title` and
 * `description` are already-localized strings handed in by the parent, and it has no empty/accessibility
 * copy. The native surface still needs accessible expand/collapse affordance labels (the web relies on the
 * DOM `role="button"` + `aria-expanded`, which TalkBack has no automatic equivalent for) and a friendly
 * empty-body hint. As with the sibling CodeBlock surface, that microcopy resolves through the i18n facade
 * by-name and falls back to these English defaults when no catalog entry exists — so this artifact stays
 * within its allowed-files scope (it adds no new catalog keys) while remaining fully localizable later.
 */
object AccordionSectionDefaults {
    /** Action label announced for the toggle while collapsed (TalkBack: "double tap to Expand"). */
    const val EXPAND_ACTION: String = "Expand"

    /** Action label announced for the toggle while expanded (TalkBack: "double tap to Collapse"). */
    const val COLLAPSE_ACTION: String = "Collapse"

    /** State description announced while expanded — the native equivalent of `aria-expanded="true"`. */
    const val EXPANDED_STATE: String = "Expanded"

    /** State description announced while collapsed — the native equivalent of `aria-expanded="false"`. */
    const val COLLAPSED_STATE: String = "Collapsed"

    /** Friendly empty-body hint (no catalog entry ⇒ this English default is used). */
    const val EMPTY_HINT: String = "Nothing to show"
}

/** Resource name for the expand action label (by-name; absent ⇒ [AccordionSectionDefaults.EXPAND_ACTION]). */
const val KEY_EXPAND_ACTION: String = "translation_accordionSection_expand"

/** Resource name for the collapse action label (by-name; absent ⇒ default). */
const val KEY_COLLAPSE_ACTION: String = "translation_accordionSection_collapse"

/** Resource name for the expanded state description (by-name; absent ⇒ default). */
const val KEY_EXPANDED_STATE: String = "translation_accordionSection_expanded"

/** Resource name for the collapsed state description (by-name; absent ⇒ default). */
const val KEY_COLLAPSED_STATE: String = "translation_accordionSection_collapsed"

/** Resource name for the empty-body hint (by-name; absent ⇒ [AccordionSectionDefaults.EMPTY_HINT]). */
const val KEY_EMPTY_HINT: String = "translation_accordionSection_empty"

/**
 * The localized strings the composable renders — resolved once at the render boundary (all by-name with the
 * web `t(key, default)` fallback, since the web source owns no catalog keys for this surface) and handed to
 * the stateless content as a framework-free bundle so the view stays a thin render layer. [actionLabel] and
 * [stateLabel] pick the open/closed variant so the header's accessibility affordances track the toggle.
 */
data class AccordionSectionStrings(
    val expandAction: String,
    val collapseAction: String,
    val expandedState: String,
    val collapsedState: String,
    val emptyHint: String,
) {
    /** The TalkBack action label for the toggle in its current [open] state (web `role="button"` intent). */
    fun actionLabel(open: Boolean): String = if (open) collapseAction else expandAction

    /** The TalkBack state description for the current [open] state (web `aria-expanded`). */
    fun stateLabel(open: Boolean): String = if (open) expandedState else collapsedState
}

/**
 * The reduced render state of an AccordionSection — the three mutually-exclusive, genuinely reachable
 * branches of the web component. The header is always present (web renders it unconditionally); the cases
 * differ only in the body region beneath it.
 */
sealed interface AccordionRender {
    /** open == false: header only, no body region (web `{open && (...)}` short-circuits). */
    data object Collapsed : AccordionRender

    /** open == true with a body slot: header + divider + the caller's content, faded in (web FadeIn). */
    data object ExpandedContent : AccordionRender

    /** open == true without a body slot: header + divider + a friendly empty state (never a blank box). */
    data object ExpandedEmpty : AccordionRender
}

/**
 * Pure, side-effect-free reducer — the native port of the web component's prop-and-state to render mapping.
 * Stateless so it is fully covered by the off-device unit gate.
 */
object AccordionSectionModel {
    /** Reproduces the web `setOpen(prev => !prev)` toggle (click + Enter/Space both flip the same flag). */
    fun toggle(open: Boolean): Boolean = !open

    /** The chevron rotation in degrees for the current [open] state (web `open && 'rotate-180'`). */
    fun chevronRotation(open: Boolean): Float = if (open) CHEVRON_OPEN_DEGREES else CHEVRON_CLOSED_DEGREES

    /** Whether the body region renders at all (web `{open && (...)}`). */
    fun shouldRenderBody(open: Boolean): Boolean = open

    /**
     * Classifies the surface into its [AccordionRender] state from the toggle [open] flag and whether a body
     * slot was supplied ([hasContent]). A collapsed section never inspects its body; an expanded section
     * with no body shows the friendly empty state instead of a blank box.
     */
    fun render(
        open: Boolean,
        hasContent: Boolean,
    ): AccordionRender =
        when {
            !open -> AccordionRender.Collapsed
            hasContent -> AccordionRender.ExpandedContent
            else -> AccordionRender.ExpandedEmpty
        }
}

/**
 * Reproduces i18next's `t(key, default)` against the native i18n facade: returns the [lookup] result for
 * [resourceName] when it resolves to a non-blank string, otherwise the [fallback] default. [lookup] is a
 * thin seam over the Android string catalog in production (an optional by-name resource read) and a map in
 * tests, so the resolve-or-fallback decision stays pure and unit-tested.
 */
fun resolveOptional(
    lookup: (String) -> String?,
    resourceName: String,
    fallback: String,
): String = lookup(resourceName)?.takeIf { it.isNotBlank() } ?: fallback
